### PR TITLE
Add meta func for scaled mm

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Adds a meta implementation for _scaled_mm which is required for dynamic shapes

cc @yanbing-j @vkuzo @albanD @kadeng

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -5179,6 +5179,38 @@ def meta__scaled_dot_product_efficient_backward(
...
+    torch._check(
+        self.dim() == 2 and mat2.dim() == 2,
+        lambda: "inputs must be 2D",
+    )
+    torch._check(
+        is_row_major(self.stride()),
+        lambda: "self must be row_major",
+    )
+    torch._check(
+        not is_row_major(mat2.stride()),
```
Comment (eellison) [inline torch/_meta_registrations.py:5205]: `not is_row_major` is not the same thing as `is_column_major`, right ?

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -5179,6 +5179,38 @@ def meta__scaled_dot_product_efficient_backward(
...
+        lambda: "mat2 must be col_major",
+    )
+
+    return torch.empty(
+        self.size(0), mat2.size(1), dtype=out_dtype, device=self.device
```
Comment (eellison) [inline torch/_meta_registrations.py:5210]: out_dtype needs to be taken from `self` if out_dtype is None

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -5179,6 +5179,38 @@ def meta__scaled_dot_product_efficient_backward(
...
+    self: torch.Tensor,
+    mat2: torch.Tensor,
+    bias: Optional[torch.Tensor] = None,
+    out_dtype: Optional[torch.dtype] = None,
+    scale_a: Optional[torch.Tensor] = None,
+    scale_b: Optional[torch.Tensor] = None,
+    scale_result: Optional[torch.Tensor] = None,
+    use_fast_accum: bool = False,
+):
+    def is_row_major(stride):
```
Comment (eellison) [inline torch/_meta_registrations.py:5193]: Check sizes are divisible by 16?

File: torch/_meta_registrations.py [hunk 4]
```diff
@@ -5179,6 +5179,38 @@ def meta__scaled_dot_product_efficient_backward(
...
+        return stride[0] > stride[1] and stride[1] == 1
+
+    torch._check(
```
Comment (eellison) [inline torch/_meta_registrations.py:5196]: add dtype checks on inputs ? https://github.com/pytorch/pytorch/blob/fd209543d5dfced3146d64df3af6361fc61fb74a/aten/src/ATen/native/cuda/Blas.cpp#L755C1-L760C1

**Review by eellison — APPROVED** [review]
Looks good - can you make sure there are opinfo inputs if there arent already ?

File: torch/_meta_registrations.py (hunk 1 above)
Comment (drisspg) [reply torch/_meta_registrations.py:5205]: you are very very correct, I will do, better checks

**Review by malfet — APPROVED** [review]

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting this change, but it is failing ROCm jobs https://hud.pytorch.org/pytorch/pytorch/commit/75174c379712433af1ff810b36e34573b3d2587e' -c nosignal

We are having capacity issue with ROCm runner, so ROCm jobs are only run in trunk atm.  I will add `ciflow/rocm` to run the jobs on your PR

**Comment by drisspg** [comment]
@huydhn no worries, I imagine that I just need to add one more decorator to the op info, but will wait for CI to confirm

**Comment by huydhn** [comment]
Just FYI, the binary build failure is unrelated and comes from https://github.com/pytorch/builder/pull/1578#issuecomment-1791836637, so that should be safe to bypass

**Pull request closed without merging**